packages/template-retail-react-app/app/components/product-item/product-quantity-picker.jsx

```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright (c) 2025, salesforce.com, inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+import React from 'react'
+import PropTypes from 'prop-types'
+import {FormattedMessage, useIntl} from 'react-intl'
+import {Text, VisuallyHidden} from '@salesforce/retail-react-app/app/components/shared/ui'
+import QuantityPicker from '@salesforce/retail-react-app/app/components/quantity-picker'
+
+const ProductQuantityPicker = ({
+    product,
+    onItemQuantityChange,
+    stepQuantity,
+    quantity,
+    setQuantity
+}) => {
+    const intl = useIntl()
+
+    return (
+        <>
+            <Text
+                fontSize="sm"
+                color="gray.700"
+                aria-label={intl.formatMessage(
+                    {
+                        id: 'item_variant.quantity.label',
+                        defaultMessage:
+                            'Quantity selector for {productName}. Selected quantity is {quantity}'
+                    },
+                    {
+                        quantity: product?.quantity,
+                        productName: product?.name
+                    }
+                )}
+            >
+                <FormattedMessage defaultMessage="Quantity:" id="product_item.label.quantity" />
+            </Text>
+            <QuantityPicker
+                step={stepQuantity}
+                value={quantity}
+                min={0}
+                clampValueOnBlur={false}
+                onBlur={(e) => {
+                    // Default to last known quantity if a user leaves the box with an invalid value
+                    const {value} = e.target
+
+                    if (!value) {
+                        setQuantity(product.quantity)
+                    }
+                }}
+                onChange={(stringValue, numberValue) => {
+                    // Set the Quantity of product to value of input if value number
+                    if (numberValue >= 0) {
+                        // Call handler
+                        onItemQuantityChange(numberValue).then(
+                            (isValidChange) => isValidChange && setQuantity(numberValue)
+                        )
+                    } else if (stringValue === '') {
+                        // We want to allow the use to clear the input to start a new input so here we set the quantity to '' so NAN is not displayed
+                        // User will not be able to add '' quantity to the cart due to the add to cart button enablement rules
+                        setQuantity(stringValue)
+                    }
+                }}
```

Comment:
It's great that we break this out, but can we go further to refactor the onchange logic into a function so that we don't mix presentation and logic code?